### PR TITLE
Fix bug in FieldResolver

### PR DESCRIPTION
**Pull request by bertfrees**
Hi @PaulRambags @kalaspuffar. I fixed a bug in FieldResolver. Fields containing soft hyphens apparently resulted in a mismatch between the text and attribute lengths. I have added a unit test to demonstrate the issue.

### CODE REVIEW

File: src/org/daisy/dotify/formatter/impl/page/FieldResolver.java [hunk 1]
```diff
@@ -122,7 +122,8 @@ public TextBorderStyle getBorder() {
         }
     }
 
-    private List<String> resolveField(
+    /* package private for unit testing */
+    List<String> resolveField(
```
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: It is usually bad practice to modify the accessibility because of unit tests. Can you please try to restore this to private and change the unit test so that method `renderField` is called and inside `renderField` `getAvailableForNoField` should be called which calls `resolveField`.

File: test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java [hunk 2]
```diff
@@ -77,4 +81,23 @@ public void testGetWidth_01() throws TranslatorConfigurationException {
...
+            null,
+            null
+        ).resolveField(
+            null,
+            new FieldList.Builder(Arrays.asList(new StringField("xxx\u00adxxx"))).build(),
+            new SimpleBrailleTranslator(
+                new DefaultBrailleFilter(new IdentityFilter(), "und", null, null),
+                new DefaultBrailleFinalizer(), "bypass"),
+            null
+        );
```
Comment (PaulRambags) [inline test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java:101]: What do you test here?
I see no `assert` statement

**Review by PaulRambags — CHANGES_REQUESTED** [review]

File: test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java (hunk 2 above)
Comment (bertfrees) [reply test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java:101]: Oh, right. I'm testing that the code doesn't raise an exception. But I'll add an expected result.

File: src/org/daisy/dotify/formatter/impl/page/FieldResolver.java (hunk 1 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: You have a good point. But there is something about this that bothers me. It's true that ideally we want to test functionality, not implementation details. We don't want to have to adapt or rewrite our tests every time we do a refactoring of the code. But this kind of questions the whole idea of unit testing, doesn't it?

The ability to unit test code depends on the visibility of the components that you want to unit test, and/or the ability to mock other components. Does this mean that it is not useful to unit test components if these conditions are not fulfilled? I don't think so. Not always. Let me explain.

Take for instance the FieldResolver class. It does things that don't require a separate class. The `renderField()` method renders a header or footer line and could be moved to `PageImpl`. The `getWidth()` method returns the space that a header or footer line takes up at a certain position on a page and could be moved to `PageSequenceBuilder2`.

That wouldn't be a bad change IMO. There is nothing that requires that these two methods stay together. They don't depend on each other. Encapsulation is a necessary thing to make code readable, but it doesn't justify creating objects for no reason. Encapsulation doesn't require objects. In Java it's easy to make the mistake of writing too many classes and it often makes the code harder to read. Dotify is a good example of this.

OK so let's say we've done this refactoring. Note that the `FieldResolver` class only contains the `resolveField()` method now (which makes sense by the way given the class name), and this method would be public or package private. So now we can unit test it directly. Yay!

But what keeps us from moving `resolveField()` to `PageImpl`? It is only used there. If we do this refactoring too, does it mean it is all of a sudden not useful anymore to unit test it, just because it is not visible anymore from outside `PageImpl`? Should we now find other ways to test it, adding the complexity that comes with mocking? What if that is not possible?

Sorry for the long answer 😅, but I hope this explains why I think following your best practice questions the whole idea of unit testing, and why I'm unsure which way to go.

Comment (kalaspuffar) [reply src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: My general thought is to keep complexity down and have things bundled in logical classes/packages. So if a method is only used in one context, then having it in that file and being protected or package private is just fine.

I've worked with libraries that love to hide proper functionality in private because they don't want it to be misused. I don't think there is such a thing. If something is useful outside of the context and it could be public, then keep it public.

Then again, I don't like writing code to follow a specific pattern or keep the code strict. I would rather see something using less memory or run faster than have it use a specific paradigm.

I've never understood the love for the builder pattern. For example, can't count the number of times I've written this code:

```
        DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
        Document doc = docBuilder.newDocument();
```

Why?

I could write this code:

```
Document doc = new com.sun.org.apache.xerces.internal.dom.DocumentImpl();
```

And it would do the same thing. But if the Java API writers want to change the functionality in the future, they can. Haven't changed this since version 1.1, so will probably not happen, but it could :)

**Review by kalaspuffar — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/page/FieldResolver.java (hunk 1 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: > My general thought is to keep complexity down and have things bundled in logical classes/packages. So if a method is only used in one context, then having it in that file and being protected or package private is just fine.

Makes a lot of sense to me.

> I've worked with libraries that love to hide proper functionality in private because they don't want it to be misused. I don't think there is such a thing.

I like the idea of making functionality that is possibly useful for the users of our library public (even if out of context). But some things to note:

- The package that contains public functionality should be exported (by the Java module or OSGi bundle). For example there is the AbstractEmbosserWriter class but I can't use it because it is in a private package (org.daisy.braille.utils.impl.tools.embosser). In Dotify everything is divided into "api" packages and "impl" packages. We have to assume that the user does not have access to the "impl" packages.
- "Everything public" is also not good of course. The usefulness of having public functionality needs to be balanced with readability of our code and with the risk that method signatures will change some day and break the user's code. My way of solving this is usually by explaining in javadocs or regular comments how a method is intended to be used.
- There is always the possibility of using reflection to access private methods. There are some places in the DAISY Pipeline code where I access private methods of Dotify. These hacks remind me that a change is probably required in the Dotify API.

File: test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java (hunk 2 above)
Comment (PaulRambags) [reply test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java:101]: >Oh, right. I'm testing that the code doesn't raise an exception. But I'll add an expected result.
Maybe add a comment that you test that no exception is raised?
Or maybe you can add `@Test(expected = None.class)` as in https://stackoverflow.com/a/44801486/7479415

Comment (bertfrees) [reply test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java:101]: I'll just add an assertEquals on the return value.

File: src/org/daisy/dotify/formatter/impl/page/FieldResolver.java (hunk 1 above)
Comment (PaulRambags) [reply src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: > Sorry for the long answer 😅

I think we are discussing here something like what we prefer how the software looks like. Personally I prefer many short classes above a few classes with many methods. If you have many short classes, implicitly most methods must be public otherwise they are not accessible from the other classes. In the case of a few classes with many methods, probably quite a lot of methods can be private, which makes unit testing of a single private method more difficult (which is the case here). But I will not block this pull request - so if you change the unit test e.g. with `@Test(expected = None.class)` - which indicates that no exception is thrown -  or with a comment that no exception is thrown or with an `assert` then I'll approve this PR.

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/FieldResolver.java:126]: I want to avoid creating objects for no reason. That doesn't make code more readable. On the contrary.

Note however that we may put functionality in separate classes without creating objects (instances) of those classes. I'm less opinionated about that. I do agree that splitting up long classes may improve readability in some cases.

But regarding the essence of this discussion, namely whether directly unit testing a function that is merely for internal use is a good idea (regardless of whether it is visible from outside its class): that question is still not answered for me. I'm inclined to say yes, in certain cases it is.

Putting a function in a separate class _so that_ it can be easily unit tested comes down to making it package private.

File: test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java (hunk 2 above)
Comment (bertfrees) [reply test/org/daisy/dotify/formatter/impl/page/FieldResolverTest.java:104]: I've done this.

**Review by PaulRambags — APPROVED** [review]

**Comment by kalaspuffar** [comment]
PR approved, running regression tests.

**Comment by kalaspuffar** [comment]
Regression tests passed, merging.
